### PR TITLE
Staged replacement of the use of evaluate* and get* with evaluate() and value() [2/N]

### LINKED ISSUE #643: Cleanup legacy methods in `Val`
Current, `Val` has the following methods:
```C++
  // If this Val is an integer with a direct constant value associated with it,
  // will return the value of that constant integer. If this integer has
  // defining expressions it will return a std::nullopt. Those values should be
  // infered using evaluateInt.
  std::optional<int64_t> getInt() const;

  // If this Val is a double with a direct constant value associated with it,
  // will return the value of that constant double. If this double has
  // defining expressions it will return a std::nullopt. Those values should be
  // infered using evaluateDouble.
  std::optional<double> getDouble() const;

  // If this Val is a bool with a direct constant value associated with it,
  // will return the value of that constant bool. If this bool has defining
  // expressions it will return a std::nullopt. Those values should be infered
  // using evaluateBool.
  std::optional<bool> getBool() const;

  // If this Val is a constant integer, and its history is comprised only of
  // constant values, will return the value of that constant integer. Cannot
  // make constant as expression evaluator takes non-constant Vals.
  int64_t evaluateInt();

  // If this Val is a constant double, and its history is comprised only of
  // constant values, will return the value of that constant double. Cannot
  // make constant as expression evaluator takes non-constant Vals.
  double evaluateDouble();

  // If this Val is a constant bool, and its history is comprised only of
  // constant values, will return the value of that constant bool. Cannot
  // make constant as expression evaluator takes non-constant Vals.
  bool evaluateBool();
```
This was a legacy before we have a `PolymorphicValue`. With `PolymorphicValue`, we only need a single
```C++
const PolymorphicValue& value() const;
```
and a single
```C++
PolymorphicValue evaluate();
```
we should replace all uses of the above 6 legacy methods with the two new methods.

### PR DESCRIPTION
**Pull request by protonu**
This replaces uses of evaluate(Int, Bool, Double) with evaluate which returns a PolymorphicValue.

This also replaces the uses of get(Int, Bool, Double) with value().

This PR is the second  of a number of them to fix the issue. In the final PR, the older get* and evaluate* methods will be removed.

Fixes #643

### CODE REVIEW

**Comment by protonu** [comment]
!build

File: csrc/kernel_ir.h [hunk 1]
```diff
@@ -122,7 +122,8 @@ class Predicate final : public Val {
   }
 
   bool isTrivial() const {
-    return isConst() && value_->getBool() == true;
+    return isConst() && value_->value().hasValue() &&
```
Comment (zasdfgbnm) [inline csrc/kernel_ir.h:125]: `is<bool>` already implies `hasValue`, we can remove it.

File: csrc/codegen.cpp [hunk 2]
```diff
@@ -1587,8 +1587,9 @@ class CudaKernelGenerator : private kir::ConstIrVisitor {
               MemoryType::Global;
         });
 
-    auto pred_bool = wop->hoistedPredicate()->getBool();
-    bool is_predicated = !(pred_bool.has_value() && pred_bool.value());
+    auto pred_bool = wop->hoistedPredicate()->value();
+    bool is_predicated =
```
Comment (zasdfgbnm) [inline csrc/codegen.cpp:1591]: Per our discussion offline, should we remove the `is<bool>`

**Review by zasdfgbnm — APPROVED** [review]

**Comment by zasdfgbnm** [comment]
Seems that this is causing a H100 failure:
```
[ RUN      ] TMATest.LoadCompleteTensor1D
unknown file: Failure
C++ exception with description "false INTERNAL ASSERT FAILED at "/home/gaoxiang/Fuser2/csrc/codegen.cpp":497, please report a bug with repro script to NVFuser at https://github.com/NVIDIA/Fuser/issues. Unhandled constant type: Array<nvfuser_index_t, 1, 1> 0
Exception raised from handle at /home/gaoxiang/Fuser2/csrc/codegen.cpp:497 (most recent call first):
frame #0: nvfuser::nvfCheckFail(char const*, char const*, unsigned int, std::string const&) + 0x8b (0x7fca7c420280 in /home/gaoxiang/Fuser2/build/libnvfuser_codegen.so)
frame #1: nvfuser::nvfErrorFail(char const*, char const*, unsigned int, char const*, std::string const&) + 0x5e (0x7fca7c41f79e in /home/gaoxiang/Fuser2/build/libnvfuser_codegen.so)
frame #2: <unknown function> + 0x3953c0 (0x7fca7c3953c0 in /home/gaoxiang/Fuser2/build/libnvfuser_codegen.so)
frame #3: <unknown function> + 0x388cc1 (0x7fca7c388cc1 in /home/gaoxiang/Fuser2/build/libnvfuser_codegen.so)
frame #4: <unknown function> + 0x38f2d9 (0x7fca7c38f2d9 in /home/gaoxiang/Fuser2/build/libnvfuser_codegen.so)
frame #5: <unknown function> + 0x388cc1 (0x7fca7c388cc1 in /home/gaoxiang/Fuser2/build/libnvfuser_codegen.so)
frame #6: <unknown function> + 0x394fcc (0x7fca7c394fcc in /home/gaoxiang/Fuser2/build/libnvfuser_codegen.so)
frame #7: <unknown function> + 0x388cc1 (0x7fca7c388cc1 in /home/gaoxiang/Fuser2/build/libnvfuser_codegen.so)
frame #8: <unknown function> + 0x396371 (0x7fca7c396371 in /home/gaoxiang/Fuser2/build/libnvfuser_codegen.so)
frame #9: nvfuser::kir::ConstIrVisitor::handle(nvfuser::kir::ForLoop const*) + 0x115 (0x7fca7c597bf5 in /home/gaoxiang/Fuser2/build/libnvfuser_codegen.so)
frame #10: <unknown function> + 0x38aafc (0x7fca7c38aafc in /home/gaoxiang/Fuser2/build/libnvfuser_codegen.so)
frame #11: <unknown function> + 0x39941b (0x7fca7c39941b in /home/gaoxiang/Fuser2/build/libnvfuser_codegen.so)
frame #12: nvfuser::codegen::generateCudaKernel(nvfuser::kir::Kernel const*, std::string const&) + 0x81 (0x7fca7c399cc1 in /home/gaoxiang/Fuser2/build/libnvfuser_codegen.so)
frame #13: nvfuser::FusionExecutor::compileFusion(nvfuser::Fusion*, nvfuser::KernelArgumentHolder const&, nvfuser::LaunchParams const&, nvfuser::CompileParams) + 0x670 (0x7fca7c431460 in /home/gaoxiang/Fuser2/build/libnvfuser_codegen.so)
frame #14: <unknown function> + 0x29ebff (0x5642a6e9ebff in ./bin/nvfuser_tests)
frame #15: <unknown function> + 0x6da36b (0x5642a72da36b in ./bin/nvfuser_tests)
frame #16: <unknown function> + 0x84f2b7 (0x5642a744f2b7 in ./bin/nvfuser_tests)
frame #17: <unknown function> + 0x83cd3d (0x5642a743cd3d in ./bin/nvfuser_tests)
frame #18: <unknown function> + 0x83cf0d (0x5642a743cf0d in ./bin/nvfuser_tests)
frame #19: <unknown function> + 0x83d0e0 (0x5642a743d0e0 in ./bin/nvfuser_tests)
frame #20: <unknown function> + 0x8409dd (0x5642a74409dd in ./bin/nvfuser_tests)
frame #21: <unknown function> + 0x84f947 (0x5642a744f947 in ./bin/nvfuser_tests)
frame #22: <unknown function> + 0x83d312 (0x5642a743d312 in ./bin/nvfuser_tests)
frame #23: <unknown function> + 0x26e309 (0x5642a6e6e309 in ./bin/nvfuser_tests)
frame #24: <unknown function> + 0x27cd0 (0x7fca1b645cd0 in /usr/lib/libc.so.6)
frame #25: __libc_start_main + 0x8a (0x7fca1b645d8a in /usr/lib/libc.so.6)
frame #26: <unknown function> + 0x263e75 (0x5642a6e63e75 in ./bin/nvfuser_tests)
" thrown in the test body.
To reproduce: NVFUSER_TEST_RANDOM_SEED=1698420749 NVFUSER_TEST_ATEN_RANDOM_SEED=0 nvfuser_tests --gtest_filter='TMATest.LoadCompleteTensor1D'
[  FAILED  ] TMATest.LoadCompleteTensor1D (106 ms)
```
Looks like this should be extended:
https://github.com/NVIDIA/Fuser/blob/5a22d44ee8a4d8a8a6dd2ce63016552be61fbc43/csrc/codegen.cpp#L469-L498

**Pull request merged**